### PR TITLE
chore: split/restructure gateway

### PR DESCRIPTION
**Pull request by tvolk131**

### CODE REVIEW

File: gateway/ln-gateway/src/lib.rs [hunk 1]
```diff
@@ -1277,7 +1270,13 @@ impl Gateway {
             }
 
             for (federation_id, federation_config) in federations {
-                if let Some(client) = self.clients.read().await.get(federation_id) {
+                if let Some(client) = self
+                    .federation_manager
+                    .clients
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lib.rs:1196]: Love the idea of a `FederationManager` and a `LightningManager`. Can we abstract away some of these operations so that their internal fields are not exposed? Something like `self.federation_manager.get_client(federation_id)`? (not sure if you did this in a separate commit or are planning on a separate PR

File: gateway/ln-gateway/src/federation_manager.rs [hunk 2]
```diff
@@ -61,4 +64,41 @@ impl FederationManager {
             next_scid: Arc::new(Mutex::new(INITIAL_SCID)),
         }
     }
+
+    pub async fn add_client(
+        &self,
+        scid: u64,
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/federation_manager.rs:71]: this shouldnt need the scid as a parameter, the `FederationManager` should manage it itself right?

File: gateway/ln-gateway/src/federation_manager.rs [hunk 3]
```diff
@@ -101,4 +102,18 @@ impl FederationManager {
...
+        &self,
+        short_channel_id: u64,
+    ) -> Option<Spanned<ClientHandleArc>> {
+        let scid_to_feds = self.scid_to_federation.read().await;
+        let clients = self.clients.read().await;
+
+        let federation_id = scid_to_feds.get(&short_channel_id)?;
+        // TODO(tvolk131): Cloning the client here could cause issues with client
+        // shutdown (see `remove_client` above). Perhaps this function should take a
+        // lambda and pass it into `client.with_sync`.
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/federation_manager.rs:116]: Sorry, kind of tough to tell with everything moving around. Was this a problem before?

File: gateway/ln-gateway/src/federation_manager.rs [hunk 4]
```diff
@@ -116,4 +116,20 @@ impl FederationManager {
         // lambda and pass it into `client.with_sync`.
         clients.get(federation_id).cloned()
     }
+
+    // TODO(tvolk131): Optimize this function by adding a reverse map from
+    // federation_id to scid.
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/federation_manager.rs:121]: Might not be worth it, this map is likely to be small so I don't think performance is an issue here.

File: gateway/ln-gateway/src/lightning_manager.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+///
+///    Initializing -- begin intercepting HTLCs --> Connected
+///    Initializing -- gateway needs config --> Configuring
+///    Configuring -- configuration set --> Connected
+///    Connected -- load federation clients --> Running
+///    Running -- disconnected from lightning node --> Disconnected
+///    Disconnected -- re-established lightning connection --> Connected
+/// ```
+#[derive(Clone, Debug)]
+pub enum GatewayState {
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning_manager.rs:117]: Hmm I don't think I agree that this should be managed by `LightningManager`. The `Initializing` and `Configuring` steps have nothing to do with the Lightning node.

File: gateway/ln-gateway/src/lightning_manager.rs [hunk 6]
```diff
@@ -25,6 +27,37 @@ impl std::fmt::Debug for LightningManager {
...
+        self.set_state(GatewayState::Connected).await;
+
+        Ok((stream, ln_client))
+    }
+
+    pub async fn get_state(&self) -> GatewayState {
+        self.state.read().await.clone()
+    }
+
+    pub async fn get_state_string(&self) -> String {
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning_manager.rs:95]: eh, I dont think this is necessary, its just as ergonomic to call `to_string()` after `get_state`

File: gateway/ln-gateway/src/lightning_manager.rs [hunk 7]
```diff
@@ -31,25 +36,46 @@ impl LightningManager {
...
         }
     }
 
     pub async fn connect_route_htlcs(
         &self,
-        task_group: &mut TaskGroup,
+        task_group: TaskGroup,
     ) -> Result<(RouteHtlcStream, Arc<dyn ILnRpcClient>), LightningRpcError> {
+        // If the gateway is already connected, stop the current HTLC interception.
+        self.stop_route_htlcs().await;
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning_manager.rs:48]: I don't think we did this before. Did you find that this was necessary?

File: gateway/ln-gateway/src/lightning_manager.rs [hunk 8]
```diff
@@ -31,25 +36,46 @@ impl LightningManager {
...
     }
 
+    /// Sets the gateway state to `Disconnected` and shuts down the task that is
+    /// listening for intercepted HTLCs if it is running.
+    pub async fn disconnect_stop_route_htlcs(&self) {
+        self.set_state(GatewayState::Disconnected).await;
+        self.stop_route_htlcs().await;
+    }
+
+    async fn stop_route_htlcs(&self) {
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning_manager.rs:71]: Add function comment

File: gateway/ln-gateway/src/lightning_manager.rs [hunk 9]
```diff
@@ -16,6 +17,10 @@ pub struct LightningManager {
 
     /// The current state of the Gateway.
     state: Arc<RwLock<GatewayState>>,
+
+    /// Task group for managing HTLC interception.
+    /// Is `Some` when the gateway is intercepting HTLCs, `None` otherwise.
+    htlc_task_group_or: Arc<RwLock<Option<TaskGroup>>>,
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning_manager.rs:23]: The `Option` here is pretty ugly. Why is it necessary? We shouldn't need to recreate the `TaskGroup` anytime we disconnect from the Lightning node. Can we do something similar to what we do now, where we just pass in the `TaskGroup` as a parameter to a function, instead of storing it as a member variable here? e.g. `lightning_manager.listen_for_htlcs(task_group).await` ?

File: gateway/ln-gateway/src/lib.rs (hunk 1 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/lib.rs:1196]: 💯 agree, the data in these structs should ideally be silo'd from the rest of the gateway

Planning this in a follow-up PR!

File: gateway/ln-gateway/src/federation_manager.rs (hunk 2 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/federation_manager.rs:71]: The reason we need it here for now is because when we restart the gateway and load the clients from the DB, they are already assigned SCIDs. After some more untangling between `Gateway` and the new `LightningManager` + `FederationManager` structs, I think we should be able to ditch this. The main focus of this PR was to keep the structure of the gateway relatively unchanged but simply separate related data into their own structs.

File: gateway/ln-gateway/src/federation_manager.rs (hunk 3 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/federation_manager.rs:116]: Yes. On master, you can search for `.clients` in [`gateway/ln-gateway/src/lib.rs`](https://github.com/fedimint/fedimint/blob/master/gateway/ln-gateway/src/lib.rs) and find where `.clone()` or `.cloned()` are called on it. I'm guessing it hasn't surfaced any issues yet, but it strikes me as bad practice and could lead to the issue described in this comment. IMO ideally we should uphold the invariant that there is never more than one reference to a client by simply not using an `Arc` to hold it.

File: gateway/ln-gateway/src/lib.rs [hunk 10]
```diff
@@ -606,49 +606,43 @@ impl Gateway {
...
+                                        info!("Got no HTLC result");
+                                    }
+                                    None
+                                })
+                                .await;
+                            if let Some(ControlFlow::Continue(())) = cf {
+                                continue;
                             }
+                        } else {
+                            info!("Got no client result");
```
[the remark this answers is not in the document]
Comment (tvolk131) [reply gateway/ln-gateway/src/lib.rs:587]: Ahhh good catch! Moved to `FederationManager::get_client_for_scid` and made the error message more descriptive.

File: gateway/ln-gateway/src/federation_manager.rs (hunk 4 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/federation_manager.rs:121]: Removed.

File: gateway/ln-gateway/src/lightning_manager.rs (hunk 5 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning_manager.rs:117]: Should we just remove the `LightningManager` struct then? That just leaves it as a shell around the `LightningBuilder`

File: gateway/ln-gateway/src/lightning_manager.rs (hunk 7 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning_manager.rs:48]: It's not strictly necessary, but I wanted the `LightningManager` to fully own its task group while providing an interface that contains as few foot-guns/idiosyncrasies as possible. `LightningManager::stop_route_htlcs` is idempotent (if it's not routing any HTLCs, it just returns without doing anything), and calling that here makes `LightningManager::connect_route_htlcs` idempotent as well (since that way we're not just tossing the old task group into the wind).

File: gateway/ln-gateway/src/lightning_manager.rs (hunk 6 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning_manager.rs:95]: Done.

File: gateway/ln-gateway/src/lightning_manager.rs (hunk 8 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning_manager.rs:71]: Done.

File: gateway/ln-gateway/src/lightning_manager.rs (hunk 9 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning_manager.rs:23]: For my clarification, are you suggesting that we change

`htlc_task_group_or: Arc<RwLock<Option<TaskGroup>>>`

to

`htlc_task_group: Arc<RwLock<TaskGroup>>`

and then add a `TaskGroup` function parameter to `LightningManager::new_initializing`?

Or are you suggesting we no longer have `LightingManager` own the HTLC task group?

**Comment by tvolk131** [comment]
Going to split this up into separate PRs for:

* `FederationManager` struct
* `LightningManager` struct
* Misc/cleanup

File: gateway/ln-gateway/src/federation_manager.rs (hunk 2 above)
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/federation_manager.rs:71]: Perhaps loading the clients from the db should also be apart of `FederationManager`

File: gateway/ln-gateway/src/lightning_manager.rs (hunk 5 above)
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lightning_manager.rs:115]: I think I still like `LightningManager` for the encapsulation of HTLC interception. It might get a little messy with the `FederationManager` though, since that needs to lookup clients. Let's do the `FederationManager` refactor first and see if it makes sense then.

File: gateway/ln-gateway/src/lightning_manager.rs (hunk 9 above)
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lightning_manager.rs:23]: I was hoping to not have `LightningManager` own the task group. Just pass it in as a parameter to the HTLC interception function.

File: gateway/ln-gateway/src/federation_manager.rs (hunk 3 above)
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/federation_manager.rs:116]: I think we had some sort of reference counting in `ClientArc` that solved this issue. Maybe @dpc or @maan2003 remembers. I think if we don't use an `Arc` it would prevent us from doing operations in parallel on the same client (maybe that's ok), e.g lookup the balance via the REST API and initiate a payment
